src/engines/ioda/src/ioda/Engines/ObsStore/Attributes.cpp

```diff
@@ -55,14 +55,36 @@ std::shared_ptr<Attribute> Attribute::write(gsl::span<const char> data,
   }
   Selection m_select(start, npoints);
   Selection f_select(start, npoints);
-  attr_data_->write(data, m_select, f_select, isFill);
+
+  if (dtype == *dtype_) {
+    attr_data_->write(data, m_select, f_select, isFill);
+  } else {
+    // Convert the ObsStore data type into an HDF5 type equivalent.
+    // Then, convert the data to the requested type and return.
+    auto internal_type = dtype_->getHDF5Type();
```

Comment:
I believe these can be `const`, as can the following `size_t` variables.

Reply:
Done in https://github.com/JCSDA-internal/ioda/pull/1228